staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -92,11 +92,17 @@ func NewAvailableConditionController(
 		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "AvailableConditionController"),
 	}
 
-	apiServiceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    c.addAPIService,
-		UpdateFunc: c.updateAPIService,
-		DeleteFunc: c.deleteAPIService,
-	})
+	// resync on this one because it is low cardinality and rechecking the actual discovery
+	// allows us to detect health in a more timely fashion when network connectivity to
+	// nodes is snipped, but the network still attempts to route there.  See
+	// https://github.com/openshift/origin/issues/17159#issuecomment-341798063
+	apiServiceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addAPIService,
+			UpdateFunc: c.updateAPIService,
+			DeleteFunc: c.deleteAPIService,
+		},
+		30*time.Second)
```

Comment:
How many thread does the controller have? Each will be blocked for 30 seconds in the worst-case.

Reply:
> How many threads does the controller have? Each will be blocked for 30 seconds in the worst-case.

Each will be blocked at most 6 seconds, right?  We have a time.After block.  I've updated the code to start 5. 

We don't back up infinitely since the queue.Add is fair (first in, first out) and eliminates duplicates.  I can make it more if you like.

Reply:
Sounds good now. If we handle a handful external apiservers, it's ok.